File: LeetCode/Arrays and Strings/Subarray Product Less Than K/JavaScript Solution/Subarray Product Less Than K.js
var numSubarrayProductLessThanK = function(nums, k) {
    if (k <= 1) return 0;
    var prod = 1, ans = 0; 
    for(let i = 0, j = 0; j < nums.length; j++) {
       prod *= nums[j];
        while(prod >= k) {
            prod /= nums[i++];
        }
        ans += j - i + 1;
    }
    return ans;
};

console.log(numSubarrayProductLessThanK([1, 1, 1, 1], 1));

File: LeetCode/Arrays and Strings/Subarray Product Less Than K/C++ Solution/subarray_product_less_than_k.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    int numSubarrayProductLessThanK(vector<int>& nums, int k) {
       int prod = 1, count = 0; 

        for(int i = 0, j = 0; j < nums.size(); j++) {
            prod *= nums[j];
            
            while(prod >= k && i < j) {
                prod /= nums[i++]; //removing first number to slide the window
            }
            
            if(prod < k) {
                int len = j - i + 1;
                count += len;
            }
        }
        return count;
    }
};

int main() {
    Solution s;
    std::vector<int> nums {10, 5, 2, 6};
    int k = 100;
    cout<<s.numSubarrayProductLessThanK(nums, k)<<endl;

    return 0;

}